clu/clu.g4

```diff
@@ -22,201 +22,291 @@
+   ;
+
+field_spec_list
+   : (field_spec (',' field_spec)*)
+   ;
+
+field_spec
+   : name_list ':' type_spec
+   ;
+
+statement
+   : decl
+   | idn ':' type_spec ':=' expression
+   | decl_list ':=' invocation
+   | idn_list ':=' (invocation | expression_list)
+   | primary '.' name ':=' expression
+   | invocation
+   | 'while' expression 'do' body 'end'
+   | 'for' (decl_list | idn_list)? 'in' invocation 'do' body 'end'
+   | ('if' expression 'then' body ('elseif' expression 'then' body)* ('else' body)? 'end')
+   | ('tagcase' expression tag_arm* ('others' ':' body)? 'end')
+   | (('return' | 'yield') (expression_list)?)
+   | ('signal' name (expression_list)?)
+   | ('exit' name (expression_list)?)
+   | 'break'
+   | ('begin' body 'end')
+   | statement (('resignal' name_list) | ('except' when_handler* others_handler? 'end'))
+   ;
+
+tag_arm
+   : 'tag' name_list ('(' idn ':' type_spec ')')? ':' body
+   ;
+
+when_handler
+   : 'when' name_list ('(' '*' ')' | (decl_list)?) ':' body
+   ;
+
+others_handler
+   : 'others' ('(' idn ':' type_spec ')')? ':' body
+   ;
+
+body
+   : equate* statement*
+   ;
+
+expression_list
+   : expression (',' expression)*
+   ;
+
+expression
+   : primary
+   | '(' expression ')'
+   | '~' expression
+   | '-' expression
+   | expression '**' expression
+   | expression '//' expression
+   | expression '/' expression
+   | expression '*' expression
+   | expression '||' expression
+   | expression '+' expression
+   | expression '-' expression
+   | expression '<' expression
+   | expression '<=' expression
+   | expression '=' expression
+   | expression '>=' expression
+   | expression '>' expression
+   | expression '~<' expression
+   | expression '~<=' expression
+   | expression '~=' expression
+   | expression '~>=' expression
+   | expression '~>' expression
+   | expression '&' expression
+   | expression 'cand' expression
+   | expression '|' expression
+   | expression 'cor' expression
+   ;
+
+primary
+   : 'nil'
+   | 'true'
+   | 'false'
+   | int_literal
+   | real_literal
+   | string_literal
+   | idn (constant_list)?
+   | primary '.' name
+   | primary expression
+   | primary '(' expression_list ')'
+   | type_spec '$' (field_list | ('[' (((expression ':')? expression_list) | constant_list) ']'))
+   | ('force' type_spec?)
+   | (('up' | 'down') '(' expression ')')
+   ;
+
+invocation
+   : primary '(' expression_list ')'
+   ;
+
+field_list
+   : field (',' field)*
+   ;
+
+field
+   : name_list ':' expression
+   ;
+
+idn_list
+   : idn (',' idn)*
+   ;
+
+idn
+   : STRING
+   ;
+
+name_list
+   : name (',' name)*
+   ;
+
+name
+   : STRING
+   ;
+
+int_literal
+   : INT
+   ;
+
+real_literal
+   : FLOAT
+   ;
+
+string_literal
+   : STRINGLITERAL
+   ;
+
+STRINGLITERAL
+   : '"' ~ '"'* '"'
+   ;
+
+STRING
+   : [a-zA-Z] [a-zA-Z0-9_]*
+   ;
+
+INT
+   : [0-9]+
+   ;
+
+FLOAT
+   : [0-9]+ '.' [0-9]*
+   ;
+
+COMMENT
+   : '%' ~ [\r\n]* -> skip
+   ;
+
+WS
+   : [ \t\r\n] -> skip
+   ;
```

Comment:
I recommend using `HIDDEN` channels instead of using `skip` command (here and in general). Information is being lost if the last one is being used:

```suggestion
COMMENT
   : '%' ~ [\r\n]* -> channel(HIDDEN)
   ;

WS
   : [ \t\r\n] -> channel(HIDDEN)
   ;
```